single_kernel_mongo/managers/sharding.py

```diff
@@ -0,0 +1,689 @@
+            )
+            if not is_leaving:
+                raise DeferrableFailedHookChecksError
+
+        shard_has_tls, config_server_has_tls = self.tls_status()
+        match (shard_has_tls, config_server_has_tls):
+            case False, True:
+                raise DeferrableFailedHookChecksError(
+                    "Config-Server uses TLS but shard does not. Please synchronise encryption method."
+                )
+            case True, False:
+                raise DeferrableFailedHookChecksError(
+                    "Shard uses TLS but config-server does not. Please synchronise encryption method."
+                )
+            case _:
+                pass
+
+        if not self.is_ca_compatible():
+            raise DeferrableFailedHookChecksError(
+                "Shard is integrated to a different CA than the config server. Please use the same CA for all cluster components.",
+            )
+
+    def prepare_to_add_shard(self) -> None:
+        """Sets status and flags in relation data relevant to sharding."""
+        # if re-using an old shard, re-set flags.
+        self.state.unit_peer_data.drained = False
+        self.charm.status_manager.to_maintenance("Adding shard to config-server")
+
+    def synchronise_cluster_secrets(self, relation: Relation, leaving: bool = False) -> None:
+        """Retrieves secrets from config-server and updates them within the shard."""
+        try:
+            self.assert_pass_hook_checks(relation=relation, is_leaving=leaving)
+        except:
+            logger.info("Skipping relation changed event: hook checks did not pass.")
+            raise
+
+        keyfile = self.state.shard_state.keyfile
+        tls_ca = self.state.shard_state.internal_ca_secret
+
+        if keyfile is None:
+            logger.info("Waiting for secrets from config-server")
+            raise WaitingForSecretsError("Missing keyfile")
+
+        self.update_member_auth(keyfile, tls_ca)
+
+        if tls_ca is not None and self.dependent.tls_manager.is_waiting_for_both_certs():
+            logger.info("Waiting for requested certs before restarting and adding to cluster.")
+            raise WaitingForCertificatesError
+
+        if not self.dependent.mongo_manager.mongod_ready():
+            raise NotReadyError
+
+        if not self.charm.unit.is_leader():
+            return
+
+        operator_password = self.state.shard_state.operator_password
+        backup_password = self.state.shard_state.backup_password
+        if not operator_password or not backup_password:
+            raise WaitingForSecretsError("Missing operator password or backup password")
+
+        self.sync_cluster_passwords(operator_password, backup_password)
+
+        # We have updated our auth, config-server can add the shard.
+        self.data_requirer.update_relation_data(relation.id, {"auth-updated": "true"})
```

Comment:
interesting

Reply:
I'm trying to compose with the former charm while moving towards a correct use of the data interfaces that we're using (and it's hard).
I'm not fully satisfied about it but it works good enough